optimizations.qmd

```diff
@@ -1,90 +1,860 @@
 
-## Dynamic Model Loading
+When the need arises for practitioners to retrieve the original higher precision values, real values r can be recovered from quantized values through an operation known as **dequantization**. In the example above, this would mean performing the following operation on our quantized value:
 
-Explanation: Incorporating dynamic model loading strategies can be highly beneficial in optimizing the performance and efficiency of AI systems, particularly in memory-constrained environments. This section discusses the importance of techniques such as partial network evaluation and on-demand model streaming, which allow for flexible model operations, helping to conserve valuable computational and memory resources on embedded devices.
+r ̃ = S(Q(r) + Z) (~ should be on top, ignore)
 
-- Partial network evaluation
-- On-demand model streaming
-- Benefits for memory-constrained devices
+As discussed, some precision in the real value is lost by quantization. In this case, the recovered value r ̃ will not exactly match r due to the rounding operation. This is an important tradeoff to note; however, in many successful uses of quantization, the loss of precision can be negligible and the test accuracy remains high. Despite this, uniform quantization continues to be the current de-facto choice due to its simplicity and efficient mapping to hardware.
+
+#### Non-uniform Quantization
+
+Non-uniform quantization, on the other hand, does not maintain a consistent interval between quantized values. This approach might be used to allocate more possible discrete values in regions where the parameter values are more densely populated, thereby preserving more detail where it is most needed. For instance, in bell-shaped distributions of weights with long tails, a set of weights in a model predominantly lies within a certain range; thus, more quantization levels might be allocated to that range to preserve finer details, enabling us to better capture information. However, one major weakness of non-uniform quantization is that it requires dequantization before higher precision computations due to its non-uniformity, restricting its ability to accelerate computation compared to uniform quantization.
+
+Typically, a rule-based non-uniform quantization uses a logarithmic distribution of exponentially increasing steps and levels as opposed to linearly. Another popular branch lies in binary-code-based quantization where real number vectors are quantized into binary vectors with a scaling factor. Notably, there is no closed form solution for minimizing errors between the real value and non-uniformly quantized value, so most quantizations in this field rely on heuristic solutions. For instance, recent work formulates non-uniform quantization as an optimization problem where the quantization steps/levels in quantizer Q are adjusted to minimize the difference between the original tensor and quantized counterpart.
```

Comment:
I think it'd be nice to have a citation/link of the "recent work" here! Also, I think a short explanation of the equation below would be helpful (like you did with the other equations above).